airbyte-cdk/python/unit_tests/sources/streams/concurrent/test_concurrent_stream_processor.py

```diff
@@ -0,0 +1,621 @@
+from airbyte_cdk.sources.streams.concurrent.partition_reader import PartitionReader
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from airbyte_cdk.sources.streams.concurrent.partitions.types import PartitionCompleteSentinel
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+
+_STREAM_NAME = "stream"
+_ANOTHER_STREAM_NAME = "stream2"
+
+
+class TestConcurrentStreamProcessor(unittest.TestCase):
+    def setUp(self):
+        self._partition_enqueuer = Mock(spec=PartitionEnqueuer)
+        self._thread_pool_manager = Mock(spec=ThreadPoolManager)
+
+        self._an_open_partition = Mock(spec=Partition)
+        self._an_open_partition.is_closed.return_value = False
+        self._log_message = Mock(spec=LogMessage)
+        self._an_open_partition.to_slice.return_value = self._log_message
+        self._an_open_partition.stream_name.return_value = _STREAM_NAME
+
+        self._a_closed_partition = Mock(spec=Partition)
+        self._a_closed_partition.is_closed.return_value = True
+        self._a_closed_partition.stream_name.return_value = _ANOTHER_STREAM_NAME
+
+        self._logger = Mock(spec=logging.Logger)
+        self._slice_logger = Mock(spec=SliceLogger)
+        self._slice_logger.create_slice_log_message.return_value = self._log_message
+        self._message_repository = Mock(spec=MessageRepository)
+        self._message_repository.consume_queue.return_value = []
+        self._partition_reader = Mock(spec=PartitionReader)
+
+        self._stream = Mock(spec=AbstractStream)
+        self._stream.name = _STREAM_NAME
+        self._stream.as_airbyte_stream.return_value = AirbyteStream(
+            name=_STREAM_NAME,
+            json_schema={},
+            supported_sync_modes=[SyncMode.full_refresh],
+        )
+        self._another_stream = Mock(spec=AbstractStream)
+        self._another_stream.name = _ANOTHER_STREAM_NAME
+        self._another_stream.as_airbyte_stream.return_value = AirbyteStream(
+            name=_ANOTHER_STREAM_NAME,
+            json_schema={},
+            supported_sync_modes=[SyncMode.full_refresh],
+        )
+
+        self._record_data = {"id": 1, "value": "A"}
+        self._record = Mock(spec=Record)
+        self._record.stream_name = _STREAM_NAME
+        self._record.data = self._record_data
+
+    def test_handle_partition_done_no_other_streams_to_generate_partitions_for(self):
+        stream_instances_to_read_from = []
+
+        handler = ConcurrentStreamProcessor(
+            stream_instances_to_read_from,
+            self._partition_enqueuer,
+            self._thread_pool_manager,
+            self._logger,
+            self._slice_logger,
+            self._message_repository,
+            self._partition_reader,
+        )
+        handler._streams_currently_generating_partitions = {_STREAM_NAME}
```

Comment:
I feel these tests should be driven by the public interface only when possible. Could we do something like:

```
    def test_handle_partition_done_no_other_streams_to_generate_partitions_for(self):
        stream_instances_to_read_from = [self._stream]
        handler = ConcurrentStreamProcessor(
            stream_instances_to_read_from,
            self._partition_enqueuer,
            self._thread_pool_manager,
            self._logger,
            self._slice_logger,
            self._message_repository,
            self._partition_reader,
        )
        handler.start_next_partition_generator()
        handler.on_partition(self._stream_partition)
        handler._streams_currently_generating_partitions = {_STREAM_NAME}
        handler._streams_to_partitions = {_STREAM_NAME: {self._an_open_partition}}

        sentinel = PartitionGenerationCompletedSentinel(self._stream)
        messages = list(handler.on_partition_generation_completed(sentinel))

        expected_messages = []
        assert expected_messages == messages
```

My concerns are:
* When a user of this class will interact with `ConcurrentStreamProcessor`, it'll use the public interface. Hence, also using the public interface in the tests will mimic real usage
* As a tester, I need to know the exact internal steps with the dependencies between private variables. This is more involved an error-prone than calling public method
* If the implementation changes, the tests will probably break. This is not true if we use the public interface as this is the contract we try to uphold and therefore should be more stable

Reply:
agreed. done!